docs: added documentation about global state management (#1072)
vuestorefront/magento2: docs/guide/global-state-management.md

In the following example, I define a new store named `useWishlistStore`

```typescript
import { defineStore } from 'pinia';
import type { Wishlist } from '~/modules/GraphQL/types';

interface WishlistStore {
  wishlist: Wishlist[] | null
}

export const useWishlistStore = defineStore('wishlist', {
  state: (): WishlistStore => ({
    wishlist: null,
  }),
});
```

The `useWishlistStore` has one state field: `wishlist` that is initially null.

### Cosuming data from Pinia store

Consuming data from the store is really straightforward. Basically you need to import it and use similarly as composable:

```typescript
import { useWishlistStore } from '~/stores/wishlistStore'

export default {
  setup() {
    const wishlistStore = useWishlistStore()

    return {
      // you can return the whole store instance to use it in the template
      wishlistStore
    }
  },
}
```

In case you want to return only **specific fields**, you need to wrap the field with `computed()`. If you don’t do that, reactivity will break.

```javascript
import { useWishlistStore } from '~/stores/wishlistStore'

export default {
  setup() {
    const wishlistStore = useWishlistStore();

>     const wishlist = computed(() => wishlistStore.wishlist);

Reviewer: Please, check this https://pinia.vuejs.org/core-concepts/#using-the-store
There is a better pattern to achieve that I think.

Author: thanks, I changed the code example.

# Global state management

For global state management, we decided to use the [Pinia](https://pinia.vuejs.org/ssr/nuxt.html) library. We assume that only data needed in more than one place of the application should be kept globally. For example:

- store config

> - categories that are used in navigation (different components for desktop and mobile)

Reviewer 2: ```suggestion
- categories used in navigation (different components for desktop and mobile)
```
nitpick

- customer (user) data
- cart/checkout data

> Data that is used in one place in the application should be stored locally in composables (as a local state)

Reviewer 2: ```suggestion
Data that is used in one place in the application should be stored locally in composables
```
stored locally in composables as local state is as we say in Poland - buttery butter

```suggestion
Data that's not shared throughout the application should be stored locally in composables (as a local state)
```
"data that is used in one place in the application" - I think you meant data that's not shared?

## Difference between global and local state

> Sometimes you need to get some information in backend and display in on the screen. For example you get products from Magento and display it on category page. You need those products only in one place. So you need that data only locally in one place.

Reviewer 2: ```suggestion
An example of local state is when you download product data from Magento and display it . You need those products only in that one view. So you should keep that data contained locally only to that view.
```

* "sometimes you need to get some"
* "category page" feels to me more like our internal lingo, unless this is  Magento thing
* rework "locally in one place"

> On the other hand, some data should be accessible from more than one place in app. Examples:

Reviewer 2: ```suggestion
On the other hand, some data should be accessible from more than one place in the app. Examples:
```

> - cart - you need data about cart in minicart overlay, but also you need to access cart for example if you want to add product to the cart from category pager or product page.

Reviewer 2: ```suggestion
- cart - you fetch cart data so you can show its contents in the the cart sidebar,  but also need cart data when  adding product to the cart from a product category page or a single product's page.
```

> - configuration - various components can consume configuration so it’s also should be accessible globally

Reviewer 2: This is too vague to me. A good example would be useUiState which we use to show globally available components (sidebars, etc. toasts?)

Author: ok, changed.

> - navigation - menus are visible on desktop and mobile devices. They display for example top level categories, but mobile menu and desktop menu is totally different components. You can store information about categories globally and consume that information by any components

Reviewer 2: ```suggestion
- navigation - because of screen size limitations, certain UI elements use the same data, but display it in a different way. For example, when browsing the product category tree in our integration, the mobile category menu is a full page overlay, while on the desktop it's embedded in the page's top header. The mobile and desktop components are far away from each other in the application's structure, but thanks to global state they can get data from the same source. 
```

## Pinia stores provided by Vue Storefront 2 for Magento integration

Take a look at stores provided by the integration OOTB:

- **Config store** - it provides information about available Magento stores, currency, and active store configuration

> - **Category store** - it provides information about categories (that is displayed in navigation)

Reviewer 2: ```suggestion
- **Category store** - it provides information about product categories
```
specify which categories
I removed the (displayed in navigation) part since I thought it could be skipped

> - **Customer store** - it provides information about customer, including information is customer logged-in or not

Reviewer 2: ```suggestion
- **Customer store** - it provides information about the currently logged-in customer.  Initially, it contains a boolean flag that lets you know if there's a valid login session. If necessary, more detailed data (shipping, billing addresses etc.) are fetched.
```

> - **Cart store** - it provides information about cart

Reviewer 2: Could this be made more specific?

Author: ok

## How to create a Pinia store

> In following example, I define a new store named `useWishlistStore`

Reviewer 2: ```suggestion
In the following example, I define a new store named `useWishlistStore`
```

> ```javascript

Reviewer 2: ```suggestion
```typescript
```

import { defineStore } from 'pinia';
import type { Wishlist } from '~/modules/GraphQL/types';

interface WishlistStore {
  wishlist: Wishlist[] | null
}

export const useWishlistStore = defineStore('wishlist', {
  state: (): WishlistStore => ({
    wishlist: null,
  }),
});
```

The `useWishlistStore` has one state field: `wishlist` that is initially null.

### Cosuming data from Pinia store

Consuming data from the store is really straightforward. Basically you need to import it and use similarly as composable:

> ```javascript

Reviewer 2: ```suggestion
```typescript
```

import { useWishlistStore } from '~/stores/wishlistStore'

> export default {

Reviewer 2: I assume you did this for simplicity but optionally yo could add defineComponent here

Author: I would keep this simple.

  setup() {
    const wishlistStore = useWishlistStore()

    return {
      // you can return the whole store instance to use it in the template
      wishlistStore
    }
  },
}
```

> In case if you want to return only **specific fields**, you need to wrap field with computed. If you don’t do that, reactivity will break.

Reviewer 2: ```suggestion
In case you want to return only **specific fields**, you need to wrap the field with `computed()`. If you don’t do that, reactivity will break.
```

It's also possible to use storeToRefs to pluck properties

```javascript
import { useWishlistStore } from '~/stores/wishlistStore'

export default {
  setup() {
    const wishlistStore = useWishlistStore();
    const wishlist = computed(() => wishlistStore.wishlist);

    return {
      wishlist,
    }
  },
}
```

> ### Modiyfying data in Pinia store

Reviewer 2: ```suggestion
### Modifying data in Pinia store
```

To write data to the store we recommend using $patch like this:

```javascript
const { data } = await app.$vsf.$magento.api.wishlist();
  wishlistStore.$patch((state) => {

>   state.wishlist = data.data?.customer?.wishlists ?? [];

Reviewer 2: there's wrong indentation here

Author: fixed.

});
```

## How to use API in Pinia stores

> If you want to communicate with APi directly in store it’s possibly by using the $graphql nuxt plugin. To do so you need to define a new action in store, let’s define the `load` action:

Reviewer 2: ```suggestion
If you want to communicate with an API directly in store it’s possibly by using the $graphql Nuxt plugin. To do so you need to define a new action in store, let’s define the `load` action:
```

> If you want to communicate with APi directly in store it’s possibly by using the $graphql nuxt plugin. To do so you need to define a new action in store, let’s define the `load` action:

Reviewer 2: I'm not sure we want to recommend using $graphql - we have two API clients - graphql request and context.magento.api.$customQuery? Do we really want to commit to $graphql? I remember we wanted to remove it :P

Author: Actually, I wanted to rewrite graphql plugin to use api.customQuery so it will be good even if we remove graphql-request.

```javascript
import { defineStore } from 'pinia';
import type { Wishlist } from '~/modules/GraphQL/types';

interface WishlistStore {
  wishlist: Wishlist[] | null
}

export const useWishlistStore = defineStore('wishlist', {
  state: (): WishlistStore => ({
    wishlist: null,
  }),
  actions: {
    async load() {

>       // communication with API will be here

Reviewer 2: I think this code block isn't necessary, as below you show the whole example. It's simple enough below

Author: ok, removed.

    },
  },
});
```

GraphQL plugin is accessible by the `this` keyword. It provides two objects: `query` (for getting data)and `mutate` (for saving data).

They `query` and `mutate` objects provides the request method that receives GraphQL query as and argument and the mutate method receive a GraphQL mutation as an argument.

The request method returns a **Promise**.

Take a look at the example:

```javascript
import { defineStore } from 'pinia';
import type { Wishlist } from '~/modules/GraphQL/types';
import wishlistGql from "./categoryList.gql";

interface WishlistStore {
  wishlist: Wishlist[] | null
}

export const useWishlistStore = defineStore('wishlist', {
  state: (): WishlistStore => ({
    wishlist: null,
  }),
  actions: {
    async load() {
      const response = await this.$nuxt.$graphql.query.request<{ wishlist: Wishlist }>(wishlistGql);
      this.wishlist = response.wishlist;
    },
  },
});
```

The `wishlistGql` is the GraphQL query and can looks like this:

```javascript
import { gql } from 'graphql-request';

export default gql`
  query wishlist {
    customer {
      wishlists {
        id
        items_count
      }
    }
  }
`;
```

> When you want to use the load action in component you can do something like this:

Reviewer 2: ```suggestion
When you want to use the load action from the example above in a component you can do something like this:
```
